content/blog/2020-12-19-asmm/index.md

```diff
@@ -0,0 +1,119 @@
+- A pointer returned from a ```Call``` value instruction is assumed to own the memory it is returning. This is to make a clear distinction of who's job it is to free memory so that different functions don't double free the same memory at the end of their regions. This is more of an implementation limitation to make inference easier than a technical limitation.
+
+## Implementation of the analysis
+
+The analysis is broken up into 5 stages. First the Bril program is broken up into a control flow graph of basic blocks. This a straight forward phase which involves inserting extra labels which will show up in the resulting code. An inference pass is then run on each function to add in any region annotations that may not have been included. The function is given a region name if the programmer does not specify one. This is also when the program identifies the owner pointers from the borrowed pointers. From this point, regardless of user input, the control flow graph for each function is now fully annotated and consistent. Because this inference needs to be conservative, each argument that a program takes in is assumed to be a part of its own region. This is to avoid some kind of leakage bug where a pointer of short lived memory is stored in a long living array and it ends up leaving a gap after it is free.
+
+Once all of the pointer types are fully annotated, the program can be quickly checked for some of the current limitations. For instance, when you call ```alloc```, we can check that all of the exit blocks that can be reached from this block are domination by the current block. This is to ensure when we come through with the next phase, any call to ```alloc``` is valid and needs to be free at each of the exit blocks dominated by this. We go through one more time and remove all of the region annotations to bring the Bril extension back to standard Bril code.
+
+## Evaluation
+
+Surprisingly, the inference part of this tool is strong enough that all four of the current benchmarks that use the memory extension can be converted to this extension of Bril merely by removing the ```free``` calls. The inference pass is able to annotate the rest in these case, ```eigth-queens.bril, fib.bril. mat-mul.bril, and sieve.bril```. The other benchmarks in the current suite do not use the memory extension but they can still be handed to and returned from this pass with no change. Three out of four of these benchmarks have the exact same total dynamic instruction count between the current benchmark and running this bril extension benchmark through the analysis. ```Sieve.bril``` was slower but this was due to the reconstruction of the program from the control graph being suboptimal and is not a performance issue with the extension itself.
```

Comment:
Don't put the code backticks around English:

```
`eigth-queens.bril`, `fib.bril`, `mat-mul.bril`, and `sieve.bril`
```

Also, the first benchmark is misspelled. Also, don't capitalize filenames at the beginning of sentences.